packages/ses/src/module-proxy.js

```diff
@@ -0,0 +1,173 @@
+// For brevity, in this file, the term "module" without qualification means
+// "module exports namespace proxy", the type that the
+// `compartment.module(fullSpecifier)` function returns.
+//
+// Compartments need a mechanism to link a module from one compartment
+// to another.
+// The procedure is to use `compartment.module(specifier)` to obtain the module
+// from one compartment, possibly before importing or even merely loading
+// module, and threading it into the module map of another compartment.
+// For this to be possible, it is necessary to model the module exports
+// namespace as a proxy that will treat all access to those exported properties
+// as reference errors until the properties of the actual module are known.
+// This provides the mechanism for modeling the public module namespace and
+// eventually connecting it to to the internal module exports namespace object.
+
+import { makeAlias } from './module-load.js';
+
+const {
+  create,
+  freeze,
+  seal,
+  isSealed,
+  preventExtensions,
+  isExtensible,
+  defineProperties,
+  getOwnPropertyDescriptors,
+} = Object;
+// q, as in quote, for error messages.
+const q = JSON.stringify;
+
+// `createDeferredModule` creates a module and activator.
+// A `Compartment` can create a module for any
+// module specifier, regardless of whether it is loadable or executable, and
+// use that object as a token that can be fed into another compartment's module
+// map.
+// Only after the specified module has been analyzed is it possible for the
+// module namespace proxy to behave properly, so it throws exceptions until
+// after the compartment has executed the module.
+export const createDeferredModule = () => {
+  let internalModule;
+  const target = create(null);
+  return freeze({
+    activate(module) {
+      internalModule = module;
+      defineProperties(target, getOwnPropertyDescriptors(module));
+      if (!isExtensible(module)) {
+        preventExtensions(target);
+      }
+      if (isSealed(module)) {
+        seal(target);
+      }
+    },
+    module: new Proxy(target, {
+      get(_target, name, receiver) {
+        if (internalModule === undefined) {
+          throw new TypeError(
+            `Cannot get property ${q(
+              name,
+            )} of module exports namespace, the module has not yet begun executing`,
+          );
+        }
+        return Reflect.get(internalModule, name, receiver);
+      },
+      set(_target, name, value) {
+        if (internalModule === undefined) {
+          throw new TypeError(
+            `Cannot set property ${q(
+              name,
+            )} of module exports namespace, the module has not yet begun executing`,
+          );
+        }
+        if (!Reflect.set(internalModule, name, value)) {
```

Comment:
This is much more wrong. The proxy must be a non-assigning view of the `internalModule`. The `set` trap must always throw. It's good that most of the other mutator traps already always throw (`defineProperty`, `deleteProperty`, `setPrototypeOf`). But `preventExtensions` should also always throw.

Reply:
Ah, this was my escape hatch so that the module’s functor could seal in its final step. I’ll try again to make this work pre-populating configurable properties for all known exports and sealing *before* running the functor.

Reply:
Just pushed a change that addresses this. The module proxy should always throw on `set`.